### PR TITLE
Week 4 kelen

### PR DESCRIPTION
**Pull request by kel-man**

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -17,41 +17,143 @@
...
+    let output1 = arrayModifier(colors1)
+    let output2 = arrayModifier(colors2)
+    // expect(arrayModifier(colors1).length).toEqual(4) // checks if the length has been modified by -1
+    // expect(arrayModifier(colors1).sort()).toEqual(['blue', 'green', 'pink', 'yellow']) // checks that the remaining elements are all the ones after the 0th index
+    // expect(arrayModifier(colors2).length).toEqual(5)
+    // expect(arrayModifier(colors2).sort()).toEqual([ 'aquamarine', 'indigo', 'ochre', 'periwinkle', 'saffron' ])
+    expect(output1.length).toEqual(4)
+    expect(output1.sort()).toEqual(['blue', 'green', 'pink', 'yellow'])
+    expect(output2.length).toEqual(5)
+    expect(output2.sort()).toEqual([ 'aquamarine', 'indigo', 'ochre', 'periwinkle', 'saffron' ])
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:31]: Great work with the testing creativity!

File: code-challenges.test.js [hunk 2]
```diff
@@ -17,41 +17,143 @@
...
+    randomIndex = Math.floor(Math.random() * currentIndex);
+    currentIndex--;
+
+    // And swap it with the current element.
+    [array[currentIndex], array[randomIndex]] = [
+      array[randomIndex], array[currentIndex]];
+  }
+
+  return array;
+}
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:57]: Solid!

File: code-challenges.test.js [hunk 3]
```diff
@@ -17,41 +17,143 @@
...
+  while(0!== currentIndex){
+    i = Math.floor(Math.random() * currentIndex)
+    currentIndex--
+    [modifiedArray[currentIndex], modifiedArray[i]] = [modifiedArray[i], modifiedArray[currentIndex]]
+  }
+  return modifiedArray
+}
+// so at this point my function is operating properly but my test runs it twice, and somehow it is modifying the input array so my second test is failing.
+// I am not sure of why the argument is being mutated when I only use it to assign the value of my output at the beginning of my function
+// However, I will fix my test so that it assigns a variable to the output of my function call and then use that variable to assert equality
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:81]: Interesting I wonder what's goin' on here

File: code-challenges.test.js [hunk 4]
```diff
@@ -17,41 +17,143 @@
...
+
+const leastAndGreatest = nums => {
+  let outputArray = [nums[0], nums[0]]
+  nums.map(num => {
+    if(num > outputArray[1]) outputArray[1] = num
+    if(num < outputArray[0]) outputArray[0] = num
+    return null
+  })
+  return outputArray
+}
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:112]: good O^n notation

File: code-challenges.test.js [hunk 5]
```diff
@@ -17,41 +17,143 @@
...
+  }                                  // Now I can rip my logic from the first function and apply it in an identical way
+  for(let i=0; i<bigArray.length; i++){
+    if(output.includes(bigArray[i]) == false){
+      output.push(bigArray[i])
+    }
+  }
+  return output
+}
+
+// Had to look up the alternate for loop notation because initially I had written for(let arr in arrays) and it was throwing me for a loop
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:158]: lol

File: code-challenges.test.js [hunk 6]
```diff
@@ -17,41 +17,143 @@
...
+  for(let arr of arrays){
+    bigArray = [...bigArray, ...arr] // This handles combining all the arrays into a single array
+  }                                  // Now I can rip my logic from the first function and apply it in an identical way
+  for(let i=0; i<bigArray.length; i++){
+    if(output.includes(bigArray[i]) == false){
+      output.push(bigArray[i])
+    }
+  }
+  return output
+}
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:156]: Great Work!

File: code_challenges.rb [hunk 7]
```diff
@@ -8,10 +8,26 @@
...
+# if it does, I add it to my output array
 
+def array_selector words, substring
+  output = []
+  words.each do |word|
+    if word.include? substring
+      output.push word
+    end
+  end
+  output #It was returning the input before I added this line. I guess line 16 is what it considers to be the implicit return
```
Comment (WalkerTexas-coder) [inline code_challenges.rb:21]: Yup, always the last line.

File: code_challenges.rb [hunk 8]
```diff
@@ -20,15 +36,92 @@
...
+# As delete_if does not work on strings, I need to split my string first
+# I'll create a dictionary of vowels in an array so I can check if the current letter is a vowel as I iterate over my array
+# Then I return a joined string from the array
 
+def remove_vowels string
+  vowels = ['a', 'e', 'i', 'o', 'u']
+  stringArray = string.split('')
+  stringArray.delete_if{|letter| vowels.include? letter.downcase}
+  stringArray.join('')
+end
```
Comment (WalkerTexas-coder) [inline code_challenges.rb:48]: Good work here!

File: code_challenges.rb [hunk 9]
```diff
@@ -20,15 +36,92 @@
...
+    @wheels = 2
+    @current_speed = 0
+    @model = model
+  end
+
+  def pedal_faster
+    if @current_speed == 40
+      p 'You are going as fast as you are able'
+    end
+    @current_speed += 10 unless @current_speed >= 40
```
Comment (WalkerTexas-coder) [inline code_challenges.rb:97]: Great use of unless given speed of (mysterious unit) 40

File: code_challenges.rb [hunk 10]
```diff
@@ -20,15 +36,92 @@
...
+raleigh.brake
+raleigh.get_info
+5.times do
+  raleigh.pedal_faster
+  raleigh.get_info
+end
+4.times do
+  raleigh.brake
+  raleigh.get_info
+end
```
Comment (WalkerTexas-coder) [inline code_challenges.rb:127]: I appreciate the extra miles here

File: interview-questions.md [hunk 11]
```diff
@@ -3,64 +3,128 @@ Answer the following questions.
...
+  Instance variables are referenced with the @ character and are the equivalent of this.variable in a Javascript class. It must be included in the
+  `attr_accessor` to be accessed, and it can be given an initial value in the initialize. Instance variables belong to an object. Other variables are typically
+  defined and used in a method or function, while instance variables belong to specific instances of a class(or object).
 
   Researched answer:
-
+  An instance variable is used in OOP to give objects their own space to store data. If an object is a data structure that can both do things and know things, then
+  instance variables represent the "know" portion of this description, while methods are the "do." Instance variables describe the attributes of any given class. Instance
+  variables are encapsulated within a class, and cannot be accessed from outside it. This can be worked around using the `attr_reader`, `attr_accessor` methods. By default an
+  instance variable will return nil. They don't always have to be declared in the initialize method, but if you want them to have initial value then they must be created there.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:19]: Good stuff!

File: interview-questions.md [hunk 12]
```diff
@@ -3,64 +3,128 @@ Answer the following questions.
...
 
   Your answer:
+  In the context of our course, Ruby blocks are essentially loops. Whenever iteration occurs, the ensuing code is a block. There are many more ways to implement iteration in Ruby than in
+  Javascript, and they are quite versatile.
 
   Researched answer:
+  Blocks in Ruby are simply code blocks. With the do/end keywords, one can enclose groups of code/logic in a single block and have it all invoked as an individual piece of code. This is
+  useful in creating methods, iterating over data, etc. Ruby blocks are not executed at the time they are read. Rather, Ruby remembers the context in which the block is created and then
+  enters the method. The keywords are not necessary for a single line of code; instead, they can be omitted and replaced with { curly braces }. Blocks can be created with parameters or
+  without, and any block in Ruby is what would be considered a higher-order block in Javascript, in that they can accept methods as arguments.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:32]: Awesome

File: interview-questions.md [hunk 13]
```diff
@@ -3,64 +3,128 @@ Answer the following questions.
...
+  case a return would be helpful, and when using return inside Procs, which causes an exception.
 
 
 
 4. What is object-oriented programming? How is it different than functional programming?
 
   Your answer:
+  In object oriented programming(hereon referred to as OOP), everything is an object. If a method exists, it is defined within an object. If a value exists, it belongs
+  to an object. In Ruby this is made exceedingly straightforward, as it is already object oriented as a language. Every single thing in Ruby is an object. This is
+  so extensive that if we wanted to, we could create a class that inherits from String and change the behavior of its methods.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:53]: yessss!

File: interview-questions.md [hunk 14]
```diff
@@ -3,64 +3,128 @@ Answer the following questions.
...
+  encapsulation, which also makes it informationally secure. This code is also easier to maintain, as class inheritance in OOP is used heavily.
 
 
 
 5. What is the difference between a class and an object?
 
   Your answer:
+  A class is a blueprint. It in itself just describes behavior. You cannot call a method that belongs to a class. It doesn't have variable values, just describes them.
+  An object is the realization of the behavior described in a class. You use a class to create an object, and the object can then be used to access its methods and
+  values.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:69]: Excellent!

**Review by WalkerTexas-coder — APPROVED** [review]
Great work Kelen! Amazing job on the interview questions (and the looking ahead questions). Also thank you for the extra attention to pseudo code and walking through the steps you took. High quality stuff. Keep up the good work!

File: code-challenges.test.js (hunk 5 above)
Comment (kel-man) [reply code-challenges.test.js:158]: Pun absolutely intended

**Pull request merged**